docs/ad-hoc-mainframe-deployment.md

```diff
@@ -0,0 +1,129 @@
+    - zosmf.yml - zosmf static definition
+
+ - Run `zowe-api-dev zfs`. ZFS filesystem is be created and mounted to the `zosTargetDir`.
+ 
+ - Build deployment artifacts.
+   - Run Gradle build task
+   - Run Gradle packageLibs task
+ 
+ - Run `zowe-api-dev deploy`. Binaries are be deployed. 
+ 
+ - Run `zowe-api-dev config --name zos`. Configuration is deployed.
+ 
+ - Start deployed instance by running 'zowe-api-dev start --job'. JES Job is started. You will get Job ID of started job.
+
+ *At this point you can go and use the deployment. You will not have to repeat these steps (Except the zfs mount if you dispose of it)*
+
+#### Phase 2 - Operate the solution
+
+`zowe-api-dev status` - To understand the state of your job.
+
+`zowe-api-dev start --job` - Starts the deployment job. Make sure the job is not running before.
+
+`zowe-api-dev deploy` - After code change, you can rerun this command to redeploy the binaries again. Only patch is deployed for efficiency. You can use `--force` to redeploy everything.
+
+`zowe-api-dev config --name zos` - After configuration change, you can run this command to redeploy changed files. There is no force and the redeployment sometimes finds no changes, even when changes have been made. In that case, inspect `.zowe-api-dev` folder and remove the mirrored deployment files and rerun the command.
+
+`zowe zos-jobs cancel job <job id>` - Stop your running job. There is no better way at the moment. We have issue pending with zowe-api-dev.
+
+#### Phase 3 - Dispose of the deployment
+
+*Make sure the deployment is not started, stop the job if it is*
+
+`zowe-api-dev zfs --unmount --delete` - Unmount and delete zfs. *Not tested*
+
+### Flow of configuration
+
+User variables are stored in `user-zowe-api.json`. You should not need to change `zowe-api.json`. `config/zowe-api-dev/template-bpxbatch.jcl` is used to run main run script on mainframe. This sets the deployment configuration into the stdenv on mainframe.
+
+Wrapper script `config/zowe-api-dev/run-wrapper.sh` is launched on mainframe. This is where you can change things and configure the deployment. This wrapper sets environment, unpacks dependencies and runs Zowe's Apiml start script `zowe-install/src/main/resources/component-scripts/start.sh`, which in turn starts Apiml.
+
+*You can use the config deploy to upload the config changes only.*
+
+### Notes and Known issues
+
+ - The delta algorithm that zowe-api-dev is using is not perfect. Sometimes manual delete of `.zowe-api-dev/uploadedFiles` cache is needed to force deployment.
+ - Zowe CLI sometimes fails to upload large files with error: `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
+`. When retried, upload passes. Further optimization can be done to reduce size of our deployment.
+
+### Deployment size optimization
+
+*Following changes are additional to standard packaging*
+
+Code packaging into jars is enhanced. 
+
+#### Thin jars package
```

Comment:
Thin jar are a great idea. A good thing is that Zowe SMP/E packaging does not create an element for each .jar file so it should work without changes no matter how many .jar files will be in the api-layer.